### PR TITLE
Add ``to/from_format`` for Cosmology interchange

### LINKED ISSUE #11997: Methods to convert Cosmology to/from another format, e.g. dict.
Add ``to_format/from_format`` methods to Cosmology for interchanging between formats, eg. Cosmology <-> dict. 
Cosmology <-> other will underpin the ``read/write`` methods and allow users to more easily create their own read/write formats
This will use the use astropy's Unified IO + registry.

Part of #11559 
tags: #priority-high  #package-expert

### PR DESCRIPTION
**Pull request by nstarman**
Add ``to/from_format`` for Cosmology interchange
Register in “mapping” for Cosmology <-> dict

Fixes #11997
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
@mhvk, only the RTD is failing from a malformed table. Any insight? Also, the table incorrectly shows that 'mapping' format can't write nor auto-identify.

Screen Shot 2021-08-01 at 20 47 31

File: docs/cosmology/io.rst
Comment (mhvk) [inline docs/cosmology/io.rst:183]: `move_to_meta` is not defined...

Would be nice to ensure the code here is tested.  One suggestion is to explicit refer to the test file and to `cosmology.io` for examples.

**Review by mhvk — COMMENTED** [review]
@nstarman - this looks very nice! I'm still wondering a bit whether one should not just read/write mappings, etc., but happy to defer to your choice.

Some in-line comments. A slightly larger question is whether to keep the design minimal. In particular, do you really want to have a `move_to_meta` argument without actual need in the present code?

I'm not sure what's on with the table of supported formats - this may require @taldcroft!

File: astropy/cosmology/_io/__init__.py
Comment (nstarman) [reply astropy/cosmology/_io/__init__.py:9]: Let me check if that works now, but there was an issue at some point with circular imports.

File: astropy/cosmology/_io/mapping.py
Comment (nstarman) [reply astropy/cosmology/_io/mapping.py:111]: I want to figure out which ones are / are not necessary. When I was writing this, the UnifiedReadWrite + registry stuff was erroring, and it stopped when I stuck in the args, kwargs. When I figure it out, I'll probably do a push on the documentation for unified-io explaining.

File: docs/changes/cosmology/11998.feature.rst
Comment (nstarman) [reply docs/changes/cosmology/11998.feature.rst:4]: Ah. thanks. I also need to do that for the read/write PR.

File: astropy/cosmology/_io/__init__.py
Comment (nstarman) [reply astropy/cosmology/_io/__init__.py:9]: It works. 👍

File: astropy/cosmology/_io/mapping.py
Comment (nstarman) [reply astropy/cosmology/_io/mapping.py:47]: Done.

File: astropy/cosmology/_io/mapping.py
Comment (nstarman) [reply astropy/cosmology/_io/mapping.py:33]: Done. And they don't appear to be needed in this case.

File: astropy/cosmology/_io/mapping.py
Comment (nstarman) [reply astropy/cosmology/_io/mapping.py:90]: Done.

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:83]: Done!

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py:153]: Thanks. Done!

File: astropy/cosmology/_io/mapping.py
Comment (nstarman) [reply astropy/cosmology/_io/mapping.py:108]: Depends on the markdown processor. These docstring don't show up in the documentation since they are only accessed by ``to/from_format``.

Comment (nstarman) [reply astropy/cosmology/_io/mapping.py:108]: On that note, I don't see documentation for the various I/O in table. Is there discussion for a place to include the contents of the I/O registry?

File: astropy/cosmology/_io/__init__.py
Comment (nstarman) [reply astropy/cosmology/_io/__init__.py:5]: 👍 Done.

**Comment by mhvk** [comment]
@nstarman - yes, I had noticed that too. I think the registries should be separated (maybe also for `Cosmology.read/write` vs `Table.read/write`?). But probably best to discuss in a separate issue (though let's ping @taldcroft here already)

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:183]: About ``move_to_meta`` here and in the comment above... thanks! I've improved the docs to have doctest examples. While ``move_to_meta`` isn't necessary presently, it's integral to allowing 3rd party registries where cosmology objects have e.g. precision flags.  I just need to make some tests in this PR that use ``move_to_meta``.

Comment (nstarman) [reply docs/cosmology/io.rst:183]: I'm actually trying to remove all mention of ``cosmology.io`` in the narrative docs. The test examples now use ``Cosmology.to/from_format`` instead of ``to/from_mapping`` and point users to ``Cosmology.to/from_format.help()``. I would like to be able to show the details of ``help`` for each format without linking to the private function, but I don't think that's currently possible.

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:116]: See https://github.com/astropy/astropy/pull/11998#discussion_r682024736. I'm completing it a different way.

File: astropy/cosmology/tests/test_mapping.py
Comment (nstarman) [reply astropy/cosmology/tests/test_mapping.py:13]: Done!

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:205]: Related to this code block. ``tox -e build_docs`` is claiming ``WARNING: Error in "code-block" directive:
unknown option: "hide"." ``, but when I inspect the built docs everything looks good...
Curious.

**Comment by nstarman** [comment]
@mhvk @taldcroft, I don't think separating registries is a blocker for this PR, though I might add a note that using ``read`` for ``to_format`` and vice versa is a soon-to-be-deprecated bug.

**Comment by nstarman** [comment]
Was not expecting that error. Apparently ``tempfile``s are forbidden in the doctests?

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py]: Tested and it can be None! that doesn't permit auto-identification, but it's moot since ``format`` is specified.

File: docs/cosmology/io.rst
Comment (taldcroft) [inline docs/cosmology/io.rst:155]: This line is confusing. Is it a requirement that the Table / Row representation have `meta['cosmology'] = <cosmology_name>`? To put in another way, sloshing the `cosmology` into `meta` for this representation seems weird. Why can't it just be in the main params `p` as a string?

File: docs/cosmology/io.rst
Comment (taldcroft) [inline docs/cosmology/io.rst:194]: Seeing this here, the distinction between `name` and `cosmology` needs to be more clear, especially in this context since people will likelly assume that `name` is the field that defines the cosmology class. I wonder if `name` should be called something more different, perhaps `id`? Or the default for `name` should be to make it all lower case so it doesn't look like a class name?

As above, I would think that the `cosmology` field should be seen here and it should be the first column.

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py:129]: Thanks! I'm implementing ``[]`` -> ``<>`` in this file and also in ``io.rst``

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py:210]: Done!

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:72]: Done! I include an example of ``.list_formats()`` and also a table of the available built-in formats, since I note that ``.list_formats`` can include registered 3rd party formats.

File: astropy/cosmology/io/mapping.py
Comment (nstarman) [reply astropy/cosmology/io/mapping.py:23]: In previous conversation with @mhvk it was suggested to make the module ``_io`` -> ``io`` since my example code for writing a custom ``read/write/to_format/from_format`` included ``to/from_mapping``. I've since updated everything to exclusively use the publicly scoped ``Cosmology.to_format("mapping")``. Rather than making these functions prefixed with an underscore, I think I'll revert the module rename and make it explicitly private.

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:155]: Ah, yeah. It probably should be for this example. In a previous PR (now closed) I had implemented a ``table`` format where the Cosmology was stored in meta. The thinking there was that a ``table`` would allow an MCMC chain of cosmology parameters, where each iteration was a row. This is normally done with a numpy array, but that doesn't do units or metadata. As the cosmology never changes it's metadata, but the "name" column could be the index. That was too much for this example, so long story short I think you're right! Changing it now.

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py:139]: Done!

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py:121]: I like this idea, but I think when the registries are separated then there will be a more significant difference. I think I'll hold off on creating a base class for now.

File: astropy/cosmology/connect.py
Comment (nstarman) [reply astropy/cosmology/connect.py:185]: see https://github.com/astropy/astropy/pull/11998#discussion_r682850019

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst]: Done!

**Comment by nstarman** [comment]
@taldcroft, the only thing not working is the malformed table. This is generated by the IO registry, which isn't getting the table header correct. Trying to debug. Any insight?

**Comment by nstarman** [comment]
@taldcroft Finally figured it out. Subtle. Numpydoc xref links anything in the Parameters, Yields, and Returns sections. The docs table was technically being generated in a Returns sections, so ``mapping`` was being transformed to ``:term:`python:mapping`, which is the wrong the length for the RST table. I now use the same trick as NDData and force the table into a ``Notes`` section to avoid numpydoc xref.

**Comment by nstarman** [comment]
Something I'd like to tackle in a future PR is more changes to ``io.registry``. The formats table for ``Cosmology.from_format`` is being written to ``Cosmology.read`` since ``io.registry`` doesn't have access to ``__set_name__`` to introspect and find the correct attribute name.

**Comment by nstarman** [comment]
Tests failing because of #12050

File: docs/cosmology/io.rst
Comment (nstarman) [inline docs/cosmology/io.rst:312]:
```suggestion
    ...        json.dump(data, file)
```

File: docs/cosmology/io.rst
Comment (nstarman) [inline docs/cosmology/io.rst:116]:
```suggestion
to the ``Planck18`` cosmology from which it was created.
```

**Comment by nstarman** [comment]
ping @mhvk @taldcroft. I think this PR is pretty much ready. #12015 will clean up some of the registry issues, but that's for followup.

> I'm still wondering a bit whether one should not just read/write mappings, etc..

I think I want to emphasize the difference between serializing to a file and creating a "live" code object. These distinctions can be blurred, e.g. a StringIO is only kind of a file, but the normal use case should be quite different. ``read/write`` will be for files while ``to/from_format`` will be for other package object. ``mapping`` is necessary for developers to stick to public API when making 3rd party package formats interoperable with Cosmology.

> A slightly larger question is whether to keep the design minimal. In particular, do you really want to have a move_to_meta argument without actual need in the present code?

I like your suggestion of a common base class in ``connect.py``, but I want to wait for #12015 for separate registries.
The ``move_to_meta`` will be integrally important for #11570 and also 3rd-party packages. I've added some tests of ``move_to_meta``.

File: astropy/cosmology/__init__.py
Comment (nstarman) [inline astropy/cosmology/__init__.py:13]:
```suggestion
from . import io  # isort: split  # needed before 'realizations'
```

File: astropy/cosmology/connect.py
Comment (nstarman) [inline astropy/cosmology/connect.py:36]:
```suggestion
      >>> Cosmology.read.help(format='<format>')  # Get detailed help on a format
```

File: astropy/cosmology/connect.py
Comment (nstarman) [inline astropy/cosmology/connect.py:105]:
```suggestion
      >>> Cosmology.write.help(format='<format>')  # Get detailed help on format
```

File: astropy/cosmology/io/mapping.py
Comment (nstarman) [inline astropy/cosmology/io/mapping.py:9]:
```suggestion
"""  # this is shown in the docs.
```

File: astropy/cosmology/io/mapping.py
Comment (nstarman) [inline astropy/cosmology/io/mapping.py:140]:
```suggestion
    parameters as items, and the metadata as a nested dictionary.
```

**Review by taldcroft — APPROVED** [review]
@nstarman - from my perspective this looks great. Again this is more high-level / design review and noting that the PR has had a good level of scrutiny. It's a big-ish PR with new functionality and if issues come up I'm sure you'll address them.

It's up to @mhvk for final approval.

File: astropy/cosmology/connect.py
Comment (mhvk) [reply astropy/cosmology/connect.py:141]: OK, can indeed refactor later, when perhaps it is clear that there is less in common than appears to be the case now.

**Review by mhvk — APPROVED** [review]
Only some very small things left, none of which are essential (well, do edit or remove the incomplete docstring for a test function). Approving so feel free to merge after.

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:90]: Yeah, no need for it twice. I like the look of the manual one better in the docs, but it's not auto-updating.

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:161]: Nifty. Didn't know it could do that. Thanks!

File: docs/cosmology/io.rst
Comment (nstarman) [reply docs/cosmology/io.rst:185]: Yeah! that would be nice. I chose Row as the simpler example to Table, but needed to use a Table anyway! 🧐

File: astropy/cosmology/tests/test_connect.py
Comment (nstarman) [reply astropy/cosmology/tests/test_connect.py:87]: Agreed. Removing.

**Comment by nstarman** [comment]
Rebased to fix the merge conflict with ``cosmology.index.rst``.

**Comment by nstarman** [comment]
Thanks @mhvk and @taldcroft for the multiple detailed rounds of reviews! 4th time was the charm.

**Comment by nstarman** [comment]
Yikes another merge conflict. Resolving now.

**Comment by taldcroft** [comment]
:tada:

**Pull request merged**
